http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -141,14 +154,22 @@ class SamAuthProvider[F[_]: OpenTelemetryMetrics](
     ev: Ask[F, TraceId]
   ): F[List[(GoogleProject, R)]] = {
     val authHeader = Authorization(Credentials.Token(AuthScheme.Bearer, userInfo.accessToken.token))
+    val resourceGoogleProjects = resources.map(_._1).toList.toSet
+    val resourceTypes = resources.map(r => sr.resourceType(r._2)).toList.toSet
     for {
-      projectPolicies <- samDao.getResourcePolicies[ProjectSamResourceId](authHeader)
-      owningProjects = projectPolicies.collect { case (r, SamPolicyName.Owner) =>
-        r.googleProject
-      }
-      resourcePolicies <- samDao
-        .getResourcePolicies[R](authHeader)
-      res = resourcePolicies.filter { case (_, pn) => sr.policyNames.contains(pn) }
+      projectPolicies <- resourceGoogleProjects.toList.flatTraverse(resourceGoogleProject =>
```

Comment:
I think this is still making too many Sam queries -- it's calling Sam for every googleProject resource in the input list, but really it should only be making at most 3 Sam calls:

* 1 for google-project resource type
* 1 for kubernetes-app resource type (if any exist in the input list)
* 1 for kubernetes-app-shared resource type (if any exist in the input list)

This may seem minor but could have a real impact if a large `listApps` query ends up pounding Sam.

@LizBaldo really no need to address this tonight -- I can try to push a commit later/tomorrow & see this though.

Reply:
Er, this method is not apps-specific but my overall comment stands  -- it should be calling Sam for `google-project` resource type and the distinct set of resource types in the set of input resources (of type `R`).

Reply:
Actually wasn't too bad, pushed a commit

Reply:
Ohh that makes a lot of sense, I was trying to generalize and be fancy, but might as well be explicit